### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [79]:
# Dependencies and Setup
import pandas as pd
import os

# File to Load (Remember to Change These)
file_to_load = os.path.join("Resources","purchase_data.csv")


# Read Purchasing File and store into Pandas data frame
purchase_data_df = pd.read_csv(file_to_load)
purchase_data_df.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## Player Count

* Display the total number of players


In [80]:
# Check to see if there are any users with multiple purchases, indeed there are
purchase_data_df['SN'].value_counts()

Lisosia93       5
Idastidru52     4
Iral74          4
Saistyphos30    3
Silaera56       3
               ..
Saena89         1
Hailaphos89     1
Iskirra45       1
Marirrasta50    1
Yasrisu92       1
Name: SN, Length: 576, dtype: int64

In [81]:
# length of list containing unique user names to get total users.
total_players = purchase_data_df['SN'].nunique()

total_players_df = pd.DataFrame({
    'Total Players': [total_players]
})
# display total players without index number for cleaner output
total_players_df.style.hide_index()


## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [82]:
# Gather summary statistics
unique_items = len(purchase_data_df['Item Name'].unique())
avg_price = purchase_data_df['Price'].mean()
total_purchases = purchase_data_df['Purchase ID'].count()
total_revenue = purchase_data_df['Price'].sum()

# Create summary table dataframe
purchase_analysis_df = pd.DataFrame({
    'Number of Unique Items': [unique_items],
    'Average Price': [avg_price],
    'Number of Purchase':[total_purchases],
    'Total Revenue': [total_revenue]
})

# Format price and revenue data to .00 precision and include '$'.
purchase_analysis_df['Average Price'] = purchase_analysis_df['Average Price'].map("${:.2f}".format)
purchase_analysis_df['Total Revenue'] = purchase_analysis_df['Total Revenue'].map("${:,.2f}".format)

# display table and hide index
purchase_analysis_df.style.hide_index()

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [83]:
# Create gender groupby object
gender_group = purchase_data_df.groupby('Gender')

# Gather total unique male, female, and Other/Non-disclosed entries
gender_df = gender_group.agg(Total_Count = ('SN', 'nunique'))

# Calcuate the percentage of male, female, and Other/Non-disclosed compared to total players
gender_df['Percentage_of_Players'] = (gender_df['Total_Count']/purchase_data_df['SN'].nunique())*100

# format percentage of players column and set indexes to 'Gender' column
gender_df['Percentage_of_Players'] = gender_df['Percentage_of_Players'].map("{:.2f}%".format)

gender_df


Total_Count Percentage_of_Players
Gender                                                  
Female                          81                14.06%
Male                           484                84.03%
Other / Non-Disclosed           11                 1.91%

## Insight 1

* Heroes of Pymoli player base is prodominantly male gendered with 84.03% of players registering as Male. This is 25% more skewed towards male players than the overall percentage of male gamers in the United States (59% in 2020, https://www.statista.com/statistics/232383/gender-split-of-us-computer-and-video-gamers/)


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [84]:
# create groupby object that groups data by gender
purchasing_gender_group = purchase_data_df.groupby('Gender')

# Use aggregate function to create new dataframe with the necessary functions applied to the respective columns
# Gets count of the Purchace ID column, number of unique users in the SN column, mean value of the Price column, and total sum of the Price column
purchasing_gender_df = purchasing_gender_group.agg(Total_Players = ('SN', 'nunique'), Purchase_Count = ('Purchase ID', 'count'), Average_Purchase_Price = ('Price', 'mean'), Total_Purchase_Value = ('Price', 'sum'))

# Gets average purchase per person by dividing total purchase value by the total players
purchasing_gender_df['Avg_Purchase_Per_Person'] = purchasing_gender_df['Total_Purchase_Value']/df['Total_Players']

# Reformat money columns to include $ and 0.00 precision
purchasing_gender_df['Average_Purchase_Price'] = purchasing_gender_df['Average_Purchase_Price'].map("${:.2f}".format)
purchasing_gender_df['Total_Purchase_Value'] = purchasing_gender_df['Total_Purchase_Value'].map("${:,.2f}".format)
purchasing_gender_df['Avg_Purchase_Per_Person'] = purchasing_gender_df['Avg_Purchase_Per_Person'].map("${:.2f}".format)

purchasing_gender_df


Total_Players  Purchase_Count Average_Purchase_Price  \
Gender                                                                        
Female                            81             113                  $3.20   
Male                             484             652                  $3.02   
Other / Non-Disclosed             11              15                  $3.35   

                      Total_Purchase_Value Avg_Purchase_Per_Person  
Gender                                                              
Female                             $361.94                   $4.47  
Male                             $1,967.64                   $4.07  
Other / Non-Disclosed               $50.19                   $4.56

## Insight 2

* In agreement with the above statistic showing males account for 84% of all players, males account for greater than 80% of all total purchase value. However, the average purchase price and average total purchase per person is the lowest for the male gender. This indicates that individual males, on average, spend less money on the game than the other genders. Leading marketing campaigns to increase the number of other gendered people and diversifying the player base may lead to more total sales. 

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points

* Display Age Demographics Table


In [85]:
# create age bins and assign bin names
age_bins = [0,9.9,14.9,19.9,24.9,29.9,34.9,39.9,200]
bin_names = ['<10', '10-14', '15-19', '20-24', '25-29', '30-34', '35-39', '40+']

# create Age Groups column that bins the age of each purchase
purchase_data_df["Age Groups"] = pd.cut(purchase_data_df["Age"], age_bins, labels=bin_names, include_lowest=True)

# Drop all duplicate SNs to get only unique players
purchase_data_no_dups_df = purchase_data_df.drop_duplicates(subset=['SN'])

# create group object that is data sorted by Age Group
age_group = purchase_data_no_dups_df.groupby(['Age Groups'])

# create empty age demographic summary table dataframe
age_sum_table_df = pd.DataFrame({})

# append total count and percentage of players for each age group (uses total_players from beginning of notebook)
age_sum_table_df['Total Count'] = age_group['Age Groups'].count()
age_sum_table_df['Percentage of Players'] = (age_sum_table_df['Total Count']/total_players)*100

# format table for % and rounding to .00 precision.
age_sum_table_df['Percentage of Players'] = age_sum_table_df['Percentage of Players'].map("{:.2f}%".format)


age_sum_table_df

Total Count Percentage of Players
Age Groups                                   
<10                  17                 2.95%
10-14                22                 3.82%
15-19               107                18.58%
20-24               258                44.79%
25-29                77                13.37%
30-34                52                 9.03%
35-39                31                 5.38%
40+                  12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [87]:
# Create new groupby object sorted by the 'Age Groups'
age_purchase_group = purchase_data_df.groupby(['Age Groups'])

# Create DataFrame summary table with purchase count, avg purchase price, total purchase value, and unique players per age group
age_purchase_df = age_purchase_group.agg(Purchase_Count = ('Purchase ID', 'count'), Total_Players = ('SN', 'nunique'), Average_Purchase_Price = ('Price', 'mean'), Total_Purchase_Value = ('Price', 'sum'))

# append avg total purchase per unique person by dividing total purchase value by total unique players
age_purchase_df['Avg_Total_Purchase_per_Person'] = age_purchase_df['Total_Purchase_Value']/age_purchase_df['Total_Players']

# format table with $ and 0.00 precision
age_purchase_df['Avg_Total_Purchase_per_Person'] = age_purchase_df['Avg_Total_Purchase_per_Person'].map("${:.2f}".format)
age_purchase_df['Average_Purchase_Price'] = age_purchase_df['Average_Purchase_Price'].map("${:,.2f}".format)
age_purchase_df['Total_Purchase_Value'] = age_purchase_df['Total_Purchase_Value'].map("${:,.2f}".format)

age_purchase_df

Purchase_Count  Total_Players Average_Purchase_Price  \
Age Groups                                                         
<10                     23             17                  $3.35   
10-14                   28             22                  $2.96   
15-19                  136            107                  $3.04   
20-24                  365            258                  $3.05   
25-29                  101             77                  $2.90   
30-34                   73             52                  $2.93   
35-39                   41             31                  $3.60   
40+                     13             12                  $2.94   

           Total_Purchase_Value Avg_Total_Purchase_per_Person  
Age Groups                                                     
<10                      $77.13                         $4.54  
10-14                    $82.78                         $3.76  
15-19                   $412.89                         $3.86  
20-24                 $1,114.06                         $4.32  
25-29                   $293.00                         $3.81  
30-34                   $214.00                         $4.12  
35-39                   $147.67                         $4.76  
40+                      $38.24                         $3.19

## Insight 3

* Players aged 20-24 account for the largest percentage of the player base (44%) and accordingly they account for the highly total purchase value.

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [88]:
# create groupby object sorted by the 'SN'
sn_group = purchase_data_df.groupby(['SN'])

# create summary table with purchase count, average purchase price, and total purchase price per user (SN)
sn_sum_table = sn_group.agg(Purchase_Count = ('Purchase ID', 'count'), Average_Purchase_Price = ('Price', 'mean'), Total_Purchase_Value = ('Price', 'sum'))

# sort values in descending order by the total purchase value
sn_sum_table = sn_sum_table.sort_values(by='Total_Purchase_Value', ascending=False)

# format average purchase price and total purchase value columns to $ and 0.00 precision
sn_sum_table['Total_Purchase_Value'] = sn_sum_table['Total_Purchase_Value'].map("${:.2f}".format)
sn_sum_table['Average_Purchase_Price'] = sn_sum_table['Average_Purchase_Price'].map("${:.2f}".format)

# display dataframe with only the first 5 rows.
sn_sum_table.head()

Purchase_Count Average_Purchase_Price Total_Purchase_Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, average item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [89]:
# filter purchase data to Item Id, Item name, and price only. This step is unneccesary in the current implimentation
pop_items_df = purchase_data_df[['Item ID', 'Item Name', 'Price']]

# create popular items group sorted by the Item Id and the Item Name
pop_items_group = pop_items_df.groupby(['Item ID', 'Item Name'])

# create summary table with number of purchases per item, item price, and total purchase value
pop_items_sum_df = pop_items_group.agg(Purchase_Count = ('Item ID', 'count'), Item_Price = ('Price', 'mean'), Total_Purchase_Value = ('Price', 'sum'))

# sort popular items by the purchase count in descending order
pop_items_sum_df = pop_items_sum_df.sort_values(by='Purchase_Count', ascending=False)

# format item price and total purchase value to include $ and 0.00 precision
pop_items_sum_df['Total_Purchase_Value'] = pop_items_sum_df['Total_Purchase_Value'].map("${:.2f}".format)
pop_items_sum_df['Item_Price'] = pop_items_sum_df['Item_Price'].map("${:.2f}".format)

# display top 5 items
pop_items_sum_df.head()

Purchase_Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Item_Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total_Purchase_Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [90]:
# remove '$' formatting from Total Puchase value so it can be converted back to float value to be sorted
pop_items_sum_df['Total_Purchase_Value'] = pop_items_sum_df['Total_Purchase_Value'].map(lambda x: x.replace('$', '')).astype(float)

# sort table by total purchase value in descending order
pop_items_sum_df = pop_items_sum_df.sort_values(by='Total_Purchase_Value', ascending=False)

# reformat total purchase value to include $ and 0.00 precision
pop_items_sum_df['Total_Purchase_Value'] = pop_items_sum_df['Total_Purchase_Value'].map("${:.2f}".format)

# display top 5 items.
pop_items_sum_df.head()

Purchase_Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Item_Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total_Purchase_Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

## Insight 4

* The highest grossing items all have high item prices nearing the most expensive item ($4.99). One may think that lower priced items might sell in higher quantities, however this table and the previous table indicate otherwise.

* These high priced items are driving sales and should be pushed to gamers whenever possible.